fish: 3.1.2 -> 3.2.0

veehaitch opened the pull request:
<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

###### Motivation for this change

Fish 3.2.0 is (finally) planned to be released next week: https://github.com/fish-shell/fish-shell/issues/6585#issuecomment-782862498

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [x] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

veehaitch committed:
fish: 3.1.2 -> 3.2.0

functions/{type.fish,_.fish}: removed, now a builtin

macOS: disable codesigning with CMake flag instead of patching the
CMakeLists.txt.

veehaitch committed:
nixos/fish: adapt completions patch to fish 3.2.0

veehaitch committed:
fish: fix passthru test

Patched file now uses double instead of single quotes.

krobelus commented on pkgs/shells/fish/default.nix:
```diff
@@ -179,13 +179,8 @@ let
-          -e "s|which |${which}/bin/which |"               \
-          -i "$out/share/fish/functions/_.fish"
       sed -e "s|uname|${coreutils}/bin/uname|"             \
           -i "$out/share/fish/functions/__fish_pwd.fish"   \
```
Not really related to your changes but.. I've been wondering why these are necessary. Are binaries like `uname` not in PATH?

Also some of these substitutions are likely out of date.
The "awk" substitution below is missing at least `share/functions/__fish_complete_wireshark.fish`.
You should probably add some checks that fail when new uses of sed and friends are added.

Also it seems strange that you don't change the path of coreutils' `seq` which is used quite a lot in completions.

veehaitch replied:
Thanks for your comment! I guess you're right for `coreutils` binaries as the package is part of the `propagatedBuildInputs`. On these grounds, patching `uname`, `seq`, etc. seems superfluous. Maybe @cole-h can shed some light on this.

r-burns replied:
I'm just guessing, but I think referring to the absolute paths is more pure. If you built an environment with multiple packages depending on different versions of coreutils, they might pick up the wrong ones if they just pick up the first one that's available on the PATH. Setting the absolute paths here ensures they always refer to the build input coreutils at runtime.

cole-h replied:
Basically what @r-burns said.

(Note that -- from what I understand -- `propagatedBuildInputs` does not affect the runtime PATH of any binary -- it's only for packages that include `fish` as a `buildInput`.)

I've had various TODOs regarding this whole `fish` situation on my list for months now, but haven't had the time (or, more accurately, the mental bandwidth) to dig into it.

Some slightly related discussion happened here: https://github.com/NixOS/nixpkgs/pull/92229.

veehaitch replied:
> (Note that -- from what I understand -- propagatedBuildInputs does not affect the runtime PATH of any binary; it's only for packages that include fish as a buildInput.)

Yes, you're right – my bad. So, the answer to your question, @krobelus, is that those binaries are _not necesarily_ in `PATH`. It is, however, not unlikely that we forgot to patch some commands. Probably just nobody noticed as users mostly have packages like `coreutils` in their `PATH`. 

I just noted that we don't run the tests. I guess, we should change that unless there's a specific reason that we don't.

veehaitch replied:
I added a commit to run the tests. Currently, they still run with `coreutils` as `checkInputs` which is probably not what we want.

veehaitch commented:
This seems to work fine now on macOS 11.2.1 and NixOS. Unfortunately, I couldn't complete `nixpkgs-review pr 113932` as I don't have enough disk space (requires ~50GB) at the moment.

I will update this PR as soon as 3.2.0 is officially released.

veehaitch commented:
Fish 3.2.0 is out: https://github.com/fish-shell/fish-shell/releases/tag/3.2.0

I've added a commit to use the 3.2.0 release.

cole-h commented on pkgs/shells/fish/dont-codesign-on-mac.diff:
```diff
@@ -4,9 +4,9 @@ index 7d220a032..786b60e6e 100644
-   TARGET_LINK_LIBRARIES(${target} fishlib)
--  CODESIGN_ON_MAC(${target})
- ENDFUNCTION(FISH_LINK_DEPS_AND_SIGN)
+ function(FISH_LINK_DEPS_AND_SIGN target)
```
This patch can be dropped and replaced by adding `-DMAC_CODESIGN_ID=OFF` to the `cmakeFlags` (see: https://github.com/fish-shell/fish-shell/issues/6952#issuecomment-620751620).

veehaitch replied:
Thanks! Deleted the patch and added the flag.

cole-h reviewed (changes requested):
* Would you mind rebasing on the latest master and force-pushing? It's a little strange that bumping `fish` would be a mass rebuild...
* Could you please squash the "official 3.2.0" commit into `fish: 3.1.2 -> 3.2.0`?
* Could we save the `fish: execute fish tests` for another PR, so we don't have to wait for the `get rid of the coreutils input` TODO to be addressed?

Thanks!

cole-h commented on pkgs/shells/fish/default.nix:
```diff
@@ -139,13 +139,38 @@ let
-    # For more information, see: https://github.com/fish-shell/fish-shell/issues/6952
-    patches = lib.optional stdenv.isDarwin ./dont-codesign-on-mac.diff;
+    # Fix FHS paths in tests
+    postPatch = ''
```
Can this also go in 6944e359dc35bf6c0b80a358c4601d2ead852a18?

veehaitch replied:
Ups, got confused in the rebase. Fixed, thanks for noticing!

veehaitch replied:
> I just didn't want to have to wait for that `TODO` to be finished before getting this version bump in. However, it seems like your recent push removed that, so I'm fine with merging that in with this PR (assuming it works as expected). Any future test improvements can always be done in a separate PR :)

Agreed. Pushed once again but seems like the build runs fine on `aarch64-linux` and `x86_64-linux`. Also macOS works on my machine.

veehaitch commented:
>     * Would you mind rebasing on the latest master and force-pushing? It's a little strange that bumping `fish` would be a mass rebuild...
> 
>     * Could you please squash the "official 3.2.0" commit into `fish: 3.1.2 -> 3.2.0`?

Both done!

>     * Could we save the `fish: execute fish tests` for another PR, so we don't have to wait for the `get rid of the coreutils input` TODO to be addressed?

Of course, we can move the task of running the tests in its entirety to another PR but I'd actually vouch for having any tests _now_ is already an improvement.

cole-h commented:
I just didn't want to have to wait for that `TODO` to be finished before getting this version bump in. However, it seems like your recent push removed that, so I'm fine with merging that in with this PR (assuming it works as expected). Any future test improvements can always be done in a separate PR :)

veehaitch committed:
fish: execute fish tests

Still includes `coreutils` in `checkInputs`.

macOS: disables some pexpect tests which depend on pkill/pgrep
(see #103180).

veehaitch commented:
> I just didn't want to have to wait for that `TODO` to be finished before getting this version bump in. However, it seems like your recent push removed that, so I'm fine with merging that in with this PR (assuming it works as expected). Any future test improvements can always be done in a separate PR :)

Agreed. Pushed once again but seems like the build runs fine on `aarch64-linux` and `x86_64-linux`. Also macOS works on my machine.

cole-h commented:
It looks like this is a mass-rebuild because of #110474 -- it added `fish` as a `checkInput`, which means everything that depends on `python3Packages.virtualenv` also depends on `fish`. I've opened https://github.com/NixOS/nixpkgs/pull/114809 to address this. I asked if it would be possible to fast-track that to `staging-next` so that when it next gets merged into master, you can just rebase this PR and (hopefully) see a much smaller number of rebuilds.

FRidh commented:
fish seems to propagate build inputs. Why is it doing that? Packages should typically not propagate build inputs.

cole-h commented:
@FRidh No idea. It's been there as long as I've been maintaining it (and longer, according to the commit that introduced it: bf3ca4cb484cdd9d1767f5cd6ef41704e8998009). Maybe it's time to remove it? (In a separate PR, of course.)

May be related to this line from the manual: "`nix-env` recursively installs all packages listed in this file when you install a package that has it." (from the [Perl section](https://nixos.org/manual/nixpkgs/stable/#ssec-perl-packaging), admittedly) -- back when `nix-env` was actively encouraged, this would ensure a working `fish`, maybe?

tbh, there's so much cruft in this derivation, maybe it's time to (mostly) rewrite it from scratch...

veehaitch commented:
> tbh, there's so much cruft in this derivation, maybe it's time to (mostly) rewrite it from scratch...

I couldn't agree more. Also, it would be nice if the derivation would build from a Git release tag rather than a release tarball.

FRidh commented:
> May be related to this line from the manual: "nix-env recursively installs all packages listed in this file when you install a package that has it." (from the Perl section, admittedly) -- back when nix-env was actively encouraged, this would ensure a working fish, maybe?

That would be `propagatedEnvPackages`.

cole-h commented:
Just to update anybody following this PR on what the status is:

It's basically "ready to merge" except for the mass rebuild. I'm planning to wait for staging-next to merge into master and then asking @veehaitch to rebase this (which should hopefully take care of the mass rebuild), but this might be a little longer than usual (https://github.com/NixOS/nixpkgs/pull/114809#issuecomment-788725059).

Once https://github.com/NixOS/nixpkgs/pull/114452 merges, this should be good to go.

(I don't want to rebase this to `staging` because that would take even longer than `staging-next`, and merging a mass rebuild to `master` is not ideal.)

cole-h commented:
@veehaitch https://github.com/NixOS/nixpkgs/pull/114452 merged, so if you rebase on master and force-push, this should be good to go :)

veehaitch commented:
> @veehaitch #114452 merged, so if you rebase on the latest master and force-push, this should be good to go :)

Done! :)

cole-h reviewed (commented):
Much more manageable amount of rebuilds :)

cole-h reviewed (approved):
Whoops, accidentally hit Ctrl+Enter...

Meant to approve :)

(Though I'll only get around to merging when I get home in ~6 hours)

---

EDIT: Seems to work fine from brief testing:

Result of `nixpkgs-review pr 113932` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>6 packages built:</summary>
  <ul>
    <li>direnv</li>
    <li>fish</li>
    <li>fishPlugins.fishtape</li>
    <li>fishPlugins.pure</li>
    <li>vimPlugins.direnv-vim</li>
    <li>zoxide</li>
  </ul>
</details>

cole-h commented:
Thanks, @veehaitch!
